From c5f71c21bc85fafae452e43b52d10506bb1863bb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Alfonso=20S=C3=A1nchez-Beato?=
 <alfonso.sanchez-beato@canonical.com>
Date: Tue, 21 May 2019 14:58:01 +0200
Subject: [PATCH 15/17] port-serial: reopen instead of force close on port
 hangup

Do not force close the port when receiving a hangup from the serial port. It is
possible to receive it in some cases, like when pppd has not properly restored
the port settings and we are not in CLOCAL mode.  If this happens, do a port
re-open instead to make sure we have a port with good settings.

Fixes LP: #1825904. Change proposed to upstream:
https://gitlab.freedesktop.org/mobile-broadband/ModemManager/merge_requests/112
---
 src/mm-port-serial.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/mm-port-serial.c b/src/mm-port-serial.c
index acb1aa5..93e4d23 100644
--- a/src/mm-port-serial.c
+++ b/src/mm-port-serial.c
@@ -1000,11 +1000,16 @@ common_input_available (MMPortSerial *self,
 
     if (condition & G_IO_HUP) {
         device = mm_port_get_device (MM_PORT (self));
-        mm_dbg ("(%s) unexpected port hangup!", device);
+        mm_warn ("(%s) unexpected port hangup!", device);
 
         if (self->priv->response->len)
             g_byte_array_remove_range (self->priv->response, 0, self->priv->response->len);
-        port_serial_close_force (self);
+
+        /* Cancel port reopening if one is running */
+        port_serial_reopen_cancel (self);
+
+        mm_port_serial_reopen (self, 1000, NULL, NULL);
+
         return G_SOURCE_REMOVE;
     }
 
-- 
2.17.1

